### PR TITLE
Expose public structured logging interface: fxevent

### PR DESCRIPTION
**Pull request by shirchen**
As part of #733 we need to finalize and expose a logging interface for
users to implement. After some internal discussion, we decided that
instead of the `Field{string, interface{}}` approach that can only be
used for logging, we'll emit structured events.

    package fxevent

    type Logger interface{ LogEvent(Event) }
    type Event interface{ /* closed interface */ }

Each event that Fx was previously logging will now be emitted as a
structured event type that implements the Event interface.

    package fxevent

    type Provide struct{ Constructor interface{} }

This gives implementations a lot more control over what is logged and
how, and leaves us room for implementing metrics and other
observability on Fx events in the future.

We also completely replace existing internal logging implementation to
use this new interface.

Note that #733 will need to be rebased on top of this.

Ref GO-666

### CODE REVIEW

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:83]: We should name the Signal "SignalName" or include the `os.Signal` here.
In the code where we supply this, we currently do:

    sig := (<-ch).String()

Instead, if  we do the following, we'll have the os.Signal that fired.

    sig := <-ch

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:104]: This is great!

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:122]: Not a blocker for this diff but FYI: once https://github.com/uber-go/dig/pull/280 lands, we might be able to get rid of fxreflect.FuncName completely.

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:142]: Since we're doing `e := event.(type)`, we can use `e` here to reference the variable with the matched type instead of casting `event` again.
```suggestion
		l.logger.Error("failed to start", zap.Error(e.Err))
	case StopSignalEvent:
		l.logger.Info("received signal", zap.String("signal", strings.ToUpper(e.Signal)))
	case StopErrorEvent:
		l.logger.Error("failed to stop cleanly", zap.Error(e.Err))
	case StartRollbackError:
		l.logger.Error("could not rollback cleanly", zap.Error(e.Err))
	case StartErrorEvent:
		l.logger.Error("startup failed, rolling back", zap.Error(e.Err))
```

**Review by abhinav — COMMENTED** [review]
Wow, that was quick! This is looking great. I've mostly got minor comments/side-notes. 

CC @sywhang; this ties in a little with your work too. We might end up with events for #746.

**Comment by sywhang** [comment]
> We might end up with events for #746.

Yeah, I like this change so I'll hold off on #746 until this one is merged and rebase that branch using these.

File: internal/fxlog/spy.go
Comment (sywhang) [inline internal/fxlog/spy.go:45]: Looks like this is returning the original slice and not the copied one:
```suggestion
	return events
```

File: app.go
Comment (abhinav) [inline app.go:758]: Oh, I see that we have a StartErrorEvent here. We could name this RollbackEvent, since that's what we're doing next. (We're usually emitting events right we do a thing, unless it can only be done after -- like receiving an error.)

If we call this RollbackEvent, the Err parameter can be StartErr: the error that caused the rollback to happen.

File: app.go
Comment (abhinav) [inline app.go:760]: This can be RollbackErrorEvent if we name the error above RollbackEvent.

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:59]: suggestion: LifecycleHookStartEvent (since the thing you put into the lifecycle is called a Hook)

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:64]: suggestion: LifecycleHookStopEvent

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:69]: Already suggested an alternative name for this.

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:90]: suggestion: InvokeErrorEvent

File: internal/fxlog/fxlog.go
Comment (abhinav) [reply internal/fxlog/fxlog.go:76]: Oh, hold up. fx.Supply doesn't take a function. It takes a value.

```
x := T1{}
fx.Supply(x)
```

We internally turn it into a constructor so that we can call Provide. So it's equivalent to,

```
x := T1{}
fx.Provide(func() T1 { return x })
```

---

For these, we only log the types of values (as seen in the deleted code above). So for now, I would make this a TypeName field.

```suggestion
	TypeName string
```

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:97]: StartErrorEvent

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:107]: RollbackEvent with a StartErr field

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:111]: RollbackErrorEvent

File: internal/fxlog/fxlog.go
Comment (abhinav) [inline internal/fxlog/fxlog.go:134]: this doesn't match what we were doing for SupplyEvent earlier. we don't log the constructor here because the constructor is made-up. the user only provided a value, so we only log its type.

```suggestion
				zap.String("type", e.TypeName),
```

**Review by abhinav — COMMENTED** [review]
The core logic looks fine minus the supply logging mismatch.
I have naming suggestions for some of the events, though.

File: fxevent/event.go
Comment (sywhang) [inline fxevent/event.go:45]:
```suggestion
// LifecycleHookStart is emitted whenever an OnStart hook is executed
```

File: fxevent/event.go
Comment (sywhang) [inline fxevent/event.go:50]:
```suggestion
// LifecycleHookStop is emitted whenever an OnStart hook is executed
```

**Review by sywhang — APPROVED** [review]
LGTM modulo couple of nits

**Pull request merged**
